
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Patrick Cahan">
      
      
      
        <link rel="prev" href="classifier.html">
      
      
      
      <link rel="icon" href="img/logo.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.49">
    
    
      
        <title>Plotting functions - pySingleCellNet</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Noto+Sans:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Noto Sans";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="stylesheets/extra.css">
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="light-blue" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#plotting-functions" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="pySingleCellNet" class="md-header__button md-logo" aria-label="pySingleCellNet" data-md-component="logo">
      
  <img src="img/logoGlow.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            pySingleCellNet
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Plotting functions
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/CahanLab/PySingleCellNet" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="pySingleCellNet" class="md-nav__button md-logo" aria-label="pySingleCellNet" data-md-component="logo">
      
  <img src="img/logoGlow.png" alt="logo">

    </a>
    pySingleCellNet
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CahanLab/PySingleCellNet" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="install.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="notebooks/quickstart.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="notebooks/categorize.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Explore
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="training_data.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training data
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="refs.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    References
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="utils.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Utility functions
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="classifier.html" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Classifier functions
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Plotting functions
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="plotting.html" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Plotting functions
  </span>
  

      </a>
      
        

  

<nav class="md-nav md-nav--secondary" aria-label="On this page">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      On this page
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting" class="md-nav__link">
    <span class="md-ellipsis">
      plotting
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.bar_classifier_f1" class="md-nav__link">
    <span class="md-ellipsis">
      bar_classifier_f1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.bar_compare_celltype_composition" class="md-nav__link">
    <span class="md-ellipsis">
      bar_compare_celltype_composition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.heatmap_gsea" class="md-nav__link">
    <span class="md-ellipsis">
      heatmap_gsea
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.heatmap_scores" class="md-nav__link">
    <span class="md-ellipsis">
      heatmap_scores
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.make_bivariate_cmap" class="md-nav__link">
    <span class="md-ellipsis">
      make_bivariate_cmap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.scatter_genes_oneper" class="md-nav__link">
    <span class="md-ellipsis">
      scatter_genes_oneper
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.scatter_qc_adata" class="md-nav__link">
    <span class="md-ellipsis">
      scatter_qc_adata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.spatial_contours" class="md-nav__link">
    <span class="md-ellipsis">
      spatial_contours
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.spatial_two_genes" class="md-nav__link">
    <span class="md-ellipsis">
      spatial_two_genes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.stackedbar_composition" class="md-nav__link">
    <span class="md-ellipsis">
      stackedbar_composition
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.stackedbar_composition_list" class="md-nav__link">
    <span class="md-ellipsis">
      stackedbar_composition_list
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.umap_scores" class="md-nav__link">
    <span class="md-ellipsis">
      umap_scores
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pySingleCellNet.plotting.umi_counts_ranked" class="md-nav__link">
    <span class="md-ellipsis">
      umi_counts_ranked
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="plotting-functions">plotting functions</h1>
<p>Various plotting functions.</p>


<div class="doc doc-object doc-module">



<a id="pySingleCellNet.plotting"></a>
    <div class="doc doc-contents first">








  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.bar_classifier_f1" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">bar_classifier_f1</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">bar_classifier_f1</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">ground_truth</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="n">class_prediction</span><span class="o">=</span><span class="s1">&#39;SCN_class&#39;</span><span class="p">,</span> <span class="n">bar_height</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plots a bar graph of F1 scores per class based on ground truth and predicted classifications.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>Annotated data matrix.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>ground_truth</code></b>
              (<code>str</code>, default:
                  <code>&#39;celltype&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The column name in <code>adata.obs</code> containing the true class labels. Defaults to "celltype".</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>class_prediction</code></b>
              (<code>str</code>, default:
                  <code>&#39;SCN_class&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The column name in <code>adata.obs</code> containing the predicted class labels. Defaults to "SCN_class".</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
          –
          <div class="doc-md-description">
            <p>None</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/bar.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">bar_classifier_f1</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> <span class="n">ground_truth</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">class_prediction</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;SCN_class&quot;</span><span class="p">,</span> <span class="n">bar_height</span><span class="o">=</span><span class="mf">0.8</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a bar graph of F1 scores per class based on ground truth and predicted classifications.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): Annotated data matrix.</span>
<span class="sd">        ground_truth (str, optional): The column name in `adata.obs` containing the true class labels. Defaults to &quot;celltype&quot;.</span>
<span class="sd">        class_prediction (str, optional): The column name in `adata.obs` containing the predicted class labels. Defaults to &quot;SCN_class&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate F1 scores</span>
    <span class="n">fscore</span> <span class="o">=</span> <span class="n">f1_score</span><span class="p">(</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">ground_truth</span><span class="p">],</span> 
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">class_prediction</span><span class="p">],</span> 
        <span class="n">average</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
        <span class="n">labels</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">ground_truth</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
    <span class="p">)</span>

    <span class="c1"># Get category names</span>
    <span class="n">cates</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">ground_truth</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span>

    <span class="c1"># Create a DataFrame for F1 scores</span>
    <span class="n">f1_scores_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s1">&#39;Class&#39;</span><span class="p">:</span> <span class="n">cates</span><span class="p">,</span>
        <span class="s1">&#39;F1-Score&#39;</span><span class="p">:</span> <span class="n">fscore</span><span class="p">,</span>
        <span class="s1">&#39;Count&#39;</span><span class="p">:</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">ground_truth</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">cates</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="p">})</span>

    <span class="c1"># Get colors from the .uns dictionary</span>
    <span class="n">f1_scores_df</span><span class="p">[</span><span class="s1">&#39;Color&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">f1_scores_df</span><span class="p">[</span><span class="s1">&#39;Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;SCN_class_colors&#39;</span><span class="p">])</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.constrained_layout.use&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="c1"># sns.set_theme(style=&quot;whitegrid&quot;)</span>

    <span class="c1"># fig, ax = plt.subplots(layout=&quot;constrained&quot;)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>

    <span class="n">text_size</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="mi">12</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">cates</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="mi">7</span><span class="p">)</span> <span class="c1"># Adjust text size</span>
    <span class="c1"># Plot the F1 scores with colors</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">f1_scores_df</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s1">&#39;Class&#39;</span><span class="p">,</span> 
        <span class="n">y</span><span class="o">=</span><span class="s1">&#39;F1-Score&#39;</span><span class="p">,</span> 
        <span class="n">color</span><span class="o">=</span><span class="n">f1_scores_df</span><span class="p">[</span><span class="s1">&#39;Color&#39;</span><span class="p">],</span> 
        <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">width</span><span class="o">=</span><span class="n">bar_height</span>
    <span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;F1-Score&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;F1-Scores per Class&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.1</span><span class="p">)</span>  <span class="c1"># Set x-axis limits to ensure visibility of all bars</span>

    <span class="c1"># Add the number of observations per class as text within the barplot</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">fscore</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">f1_scores_df</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">],</span> <span class="n">f1_scores_df</span><span class="p">[</span><span class="s1">&#39;F1-Score&#39;</span><span class="p">])):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.03</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;n = </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;white&#39;</span> <span class="k">if</span> <span class="n">fscore</span> <span class="o">&gt;=</span> <span class="mf">0.20</span> <span class="k">else</span> <span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">text_size</span><span class="p">)</span>

    <span class="c1"># plt.show()</span>
    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.bar_compare_celltype_composition" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">bar_compare_celltype_composition</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">bar_compare_celltype_composition</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">,</span> <span class="n">celltype_col</span><span class="p">,</span> <span class="n">min_delta</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;log_ratio&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compare cell type proportions between two AnnData objects and plot either log-ratio or differences for significant changes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata1</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>First AnnData object.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>adata2</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>Second AnnData object.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>celltype_col</code></b>
              (<code>str</code>)
          –
          <div class="doc-md-description">
            <p>Column name in <code>.obs</code> indicating cell types.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>min_delta</code></b>
              (<code>float</code>)
          –
          <div class="doc-md-description">
            <p>Minimum absolute difference in percentages to include in the plot.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>colors</code></b>
              (<code>dict</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>Dictionary with cell types as keys and colors as values for the bars.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>metric</code></b>
              (<code>str</code>, default:
                  <code>&#39;log_ratio&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>"log_ratio" (default) or "difference" to specify which metric to plot.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
<b><code>None</code></b>          –
          <div class="doc-md-description">
            <p>Displays the bar plot.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/bar.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">bar_compare_celltype_composition</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">,</span> <span class="n">celltype_col</span><span class="p">,</span> <span class="n">min_delta</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;log_ratio&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compare cell type proportions between two AnnData objects and plot either log-ratio or differences for significant changes.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        adata1 (AnnData): First AnnData object.</span>
<span class="sd">        adata2 (AnnData): Second AnnData object.</span>
<span class="sd">        celltype_col (str): Column name in `.obs` indicating cell types.</span>
<span class="sd">        min_delta (float): Minimum absolute difference in percentages to include in the plot.</span>
<span class="sd">        colors (dict, optional): Dictionary with cell types as keys and colors as values for the bars.</span>
<span class="sd">        metric (str, optional): &quot;log_ratio&quot; (default) or &quot;difference&quot; to specify which metric to plot.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: Displays the bar plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Compute cell type percentages for both AnnData objects</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">compute_percentages</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">celltype_col</span><span class="p">):</span>
        <span class="n">cell_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_col</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
        <span class="k">return</span> <span class="n">cell_counts</span>

    <span class="n">percentages_adata1</span> <span class="o">=</span> <span class="n">compute_percentages</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">celltype_col</span><span class="p">)</span>
    <span class="n">percentages_adata2</span> <span class="o">=</span> <span class="n">compute_percentages</span><span class="p">(</span><span class="n">adata2</span><span class="p">,</span> <span class="n">celltype_col</span><span class="p">)</span>

    <span class="c1"># Align indices to ensure comparison</span>
    <span class="n">all_celltypes</span> <span class="o">=</span> <span class="n">percentages_adata1</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">percentages_adata2</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">percentages_adata1</span> <span class="o">=</span> <span class="n">percentages_adata1</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">all_celltypes</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">percentages_adata2</span> <span class="o">=</span> <span class="n">percentages_adata2</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">all_celltypes</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Compute the differences and log-ratio</span>
    <span class="n">differences</span> <span class="o">=</span> <span class="n">percentages_adata1</span> <span class="o">-</span> <span class="n">percentages_adata2</span>
    <span class="n">log_ratios</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">((</span><span class="n">percentages_adata1</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">percentages_adata2</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">))</span>  <span class="c1"># Avoid division by zero</span>

    <span class="c1"># Choose the metric to plot</span>
    <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s2">&quot;log_ratio&quot;</span><span class="p">:</span>
        <span class="n">plot_values</span> <span class="o">=</span> <span class="n">log_ratios</span>
        <span class="n">xlabel</span> <span class="o">=</span> <span class="s2">&quot;Log2(Percent in A / Percent in B)&quot;</span>
        <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Log2 Ratio of Cell Type Percentages&quot;</span>
    <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s2">&quot;difference&quot;</span><span class="p">:</span>
        <span class="n">plot_values</span> <span class="o">=</span> <span class="n">differences</span>
        <span class="n">xlabel</span> <span class="o">=</span> <span class="s2">&quot;Difference in Percentages (A - B)&quot;</span>
        <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Difference in Cell Type Percentages&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid metric. Choose either &#39;log_ratio&#39; or &#39;difference&#39;.&quot;</span><span class="p">)</span>

    <span class="c1"># Filter cell types by the threshold</span>
    <span class="n">significant_celltypes</span> <span class="o">=</span> <span class="n">plot_values</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">differences</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">min_delta</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>

    <span class="c1"># Prepare data for plotting</span>
    <span class="n">plot_data</span> <span class="o">=</span> <span class="n">plot_values</span><span class="p">[</span><span class="n">significant_celltypes</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span>

    <span class="c1"># Determine colors for the bars (align with sorted data)</span>
    <span class="k">if</span> <span class="n">colors</span><span class="p">:</span>
        <span class="n">bar_colors</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="n">colors</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]))</span> <span class="k">if</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">colors</span> <span class="k">else</span> <span class="s1">&#39;skyblue&#39;</span> <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">plot_data</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">bar_colors</span> <span class="o">=</span> <span class="s1">&#39;skyblue&#39;</span>

    <span class="c1"># Debugging: Log the bar colors and sorted cell types</span>
    <span class="c1"># print(&quot;Sorted Cell Types:&quot;, plot_data.index.tolist())</span>
    <span class="c1"># print(&quot;Bar Colors:&quot;, bar_colors)</span>

    <span class="c1"># Create the horizontal bar plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">plot_data</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;barh&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">bar_colors</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)</span>
    <span class="c1"># plt.axvline(0, color=&#39;gray&#39;, linestyle=&#39;--&#39;, linewidth=1)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># plt.title(title)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span>
    <span class="c1"># plt.ylabel(&quot;Cell Types&quot;)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.heatmap_gsea" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">heatmap_gsea</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">heatmap_gsea</span><span class="p">(</span><span class="n">gmat</span><span class="p">,</span> <span class="n">clean_signatures</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">clean_cells</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">column_colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">label_font_size</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">cbar_pos</span><span class="o">=</span><span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.92</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">],</span> <span class="n">dendro_ratio</span><span class="o">=</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="n">col_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">row_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Generates a heatmap with hierarchical clustering for gene set enrichment analysis (GSEA) results.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>gmat</code></b>
              (<code><span title="pandas.DataFrame">DataFrame</span></code>)
          –
          <div class="doc-md-description">
            <p>A matrix of GSEA scores with gene sets as rows and samples as columns.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>clean_signatures</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>If True, removes gene sets with zero enrichment scores across all samples. Defaults to False.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>clean_cells</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>If True, removes samples with zero enrichment scores across all gene sets. Defaults to False.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>column_colors</code></b>
              (<code><span title="pandas.Series">Series</span> or <span title="pandas.DataFrame">DataFrame</span></code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>Colors to annotate columns, typically representing sample groups. Defaults to None.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>figsize</code></b>
              (<code>tuple</code>, default:
                  <code>(8, 6)</code>
)
          –
          <div class="doc-md-description">
            <p>Figure size in inches (width, height). Defaults to (8, 6).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>label_font_size</code></b>
              (<code>int</code>, default:
                  <code>7</code>
)
          –
          <div class="doc-md-description">
            <p>Font size for axis and colorbar labels. Defaults to 7.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>cbar_pos</code></b>
              (<code>list</code>, default:
                  <code>[0.2, 0.92, 0.6, 0.02]</code>
)
          –
          <div class="doc-md-description">
            <p>Position of the colorbar [left, bottom, width, height]. Defaults to [0.2, 0.92, 0.6, 0.02] for a horizontal top placement.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>dendro_ratio</code></b>
              (<code>tuple</code>, default:
                  <code>(0.3, 0.1)</code>
)
          –
          <div class="doc-md-description">
            <p>Proportion of the figure allocated to the row and column dendrograms. Defaults to (0.3, 0.1).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>cbar_title</code></b>
              (<code>str</code>, default:
                  <code>&#39;NES&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Title of the colorbar. Defaults to 'NES'.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>col_cluster</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>If True, performs hierarchical clustering on columns. Defaults to False.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>row_cluster</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>If True, performs hierarchical clustering on rows. Defaults to False.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
          –
          <div class="doc-md-description">
            <p>None</p>
          </div>
        </li>
    </ul>


<details class="displays" open>
  <summary>Displays</summary>
  <p>A heatmap with optional hierarchical clustering and a horizontal colorbar at the top.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/heatmap.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">heatmap_gsea</span><span class="p">(</span>
    <span class="n">gmat</span><span class="p">,</span>
    <span class="n">clean_signatures</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">clean_cells</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">column_colors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
    <span class="n">label_font_size</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
    <span class="n">cbar_pos</span><span class="o">=</span><span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.92</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">],</span>  <span class="c1"># Positioned at the top</span>
    <span class="n">dendro_ratio</span><span class="o">=</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
    <span class="n">cbar_title</span><span class="o">=</span><span class="s1">&#39;NES&#39;</span><span class="p">,</span>
    <span class="n">col_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">row_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a heatmap with hierarchical clustering for gene set enrichment analysis (GSEA) results.</span>

<span class="sd">    Args:</span>
<span class="sd">        gmat (pd.DataFrame):</span>
<span class="sd">            A matrix of GSEA scores with gene sets as rows and samples as columns.</span>
<span class="sd">        clean_signatures (bool, optional):</span>
<span class="sd">            If True, removes gene sets with zero enrichment scores across all samples. Defaults to False.</span>
<span class="sd">        clean_cells (bool, optional):</span>
<span class="sd">            If True, removes samples with zero enrichment scores across all gene sets. Defaults to False.</span>
<span class="sd">        column_colors (pd.Series or pd.DataFrame, optional):</span>
<span class="sd">            Colors to annotate columns, typically representing sample groups. Defaults to None.</span>
<span class="sd">        figsize (tuple, optional):</span>
<span class="sd">            Figure size in inches (width, height). Defaults to (8, 6).</span>
<span class="sd">        label_font_size (int, optional):</span>
<span class="sd">            Font size for axis and colorbar labels. Defaults to 7.</span>
<span class="sd">        cbar_pos (list, optional):</span>
<span class="sd">            Position of the colorbar [left, bottom, width, height]. Defaults to [0.2, 0.92, 0.6, 0.02] for a horizontal top placement.</span>
<span class="sd">        dendro_ratio (tuple, optional):</span>
<span class="sd">            Proportion of the figure allocated to the row and column dendrograms. Defaults to (0.3, 0.1).</span>
<span class="sd">        cbar_title (str, optional):</span>
<span class="sd">            Title of the colorbar. Defaults to &#39;NES&#39;.</span>
<span class="sd">        col_cluster (bool, optional):</span>
<span class="sd">            If True, performs hierarchical clustering on columns. Defaults to False.</span>
<span class="sd">        row_cluster (bool, optional):</span>
<span class="sd">            If True, performs hierarchical clustering on rows. Defaults to False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>

<span class="sd">    Displays:</span>
<span class="sd">        A heatmap with optional hierarchical clustering and a horizontal colorbar at the top.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gsea_matrix</span> <span class="o">=</span> <span class="n">gmat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">clean_cells</span><span class="p">:</span>
        <span class="n">gsea_matrix</span> <span class="o">=</span> <span class="n">gsea_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gsea_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">clean_signatures</span><span class="p">:</span>
        <span class="n">gsea_matrix</span> <span class="o">=</span> <span class="n">gsea_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gsea_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span> <span class="p">:]</span>

    <span class="c1"># plt.figure(constrained_layout=True)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">gsea_matrix</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="n">Roma_20</span><span class="o">.</span><span class="n">mpl_colormap</span><span class="o">.</span><span class="n">reversed</span><span class="p">(),</span>
        <span class="n">center</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">yticklabels</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">xticklabels</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">linewidth</span><span class="o">=</span><span class="mf">.05</span><span class="p">,</span>
        <span class="n">linecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">,</span>
        <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span>
        <span class="n">dendrogram_ratio</span><span class="o">=</span><span class="n">dendro_ratio</span><span class="p">,</span>
        <span class="n">col_colors</span><span class="o">=</span><span class="n">column_colors</span><span class="p">,</span>
        <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
        <span class="n">row_cluster</span><span class="o">=</span><span class="n">row_cluster</span><span class="p">,</span>
        <span class="n">col_cluster</span><span class="o">=</span><span class="n">col_cluster</span><span class="p">,</span>
        <span class="n">cbar_pos</span><span class="o">=</span><span class="n">cbar_pos</span><span class="p">,</span>
        <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;orientation&#39;</span><span class="p">:</span> <span class="s1">&#39;horizontal&#39;</span><span class="p">}</span>
    <span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">ax_cbar</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">cbar_title</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">label_font_size</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">ax_cbar</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">labelsize</span><span class="o">=</span><span class="n">label_font_size</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">)</span>

    <span class="c1"># Adjust tick labels and heatmap appearance</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">ax_row_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">ax_col_dendrogram</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_ymajorticklabels</span><span class="p">(),</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">label_font_size</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">ax_heatmap</span><span class="o">.</span><span class="n">get_xmajorticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">,</span> <span class="n">rotation_mode</span><span class="o">=</span><span class="s2">&quot;anchor&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">label_font_size</span><span class="p">)</span>

    <span class="c1"># plt.subplots_adjust(top=0.85)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.heatmap_scores" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">heatmap_scores</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">heatmap_scores</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">obsm_name</span><span class="o">=</span><span class="s1">&#39;SCN_score&#39;</span><span class="p">,</span> <span class="n">order_by</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figure_subplot_bottom</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plots a heatmap of single cell scores, grouping cells according to a specified .obs column and optionally ordering within each group.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>An AnnData object containing the single cell data.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>groupby</code></b>
              (<code>str</code>)
          –
          <div class="doc-md-description">
            <p>The name of the column in .obs used for grouping cells in the heatmap.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>vmin</code></b>
              (<code>float</code>, default:
                  <code>0</code>
)
          –
          <div class="doc-md-description">
            <p>Minimum value for color scaling. Defaults to 0.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>vmax</code></b>
              (<code>float</code>, default:
                  <code>1</code>
)
          –
          <div class="doc-md-description">
            <p>Maximum value for color scaling. Defaults to 1.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>obsm_name</code></b>
              (<code>str</code>, default:
                  <code>&#39;SCN_score&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The key in .obsm to retrieve the matrix for plotting. Defaults to 'SCN_score'.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>order_by</code></b>
              (<code>str</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>The name of the column in .obs used for ordering cells within each group. Defaults to None.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
<b><code>None</code></b>          –
          <div class="doc-md-description">
            <p>The function plots a heatmap and does not return any value.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/heatmap.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">heatmap_scores</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> 
    <span class="n">groupby</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> 
    <span class="n">vmin</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> 
    <span class="n">vmax</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> 
    <span class="n">obsm_name</span><span class="o">=</span><span class="s1">&#39;SCN_score&#39;</span><span class="p">,</span> 
    <span class="n">order_by</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">figure_subplot_bottom</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.4</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a heatmap of single cell scores, grouping cells according to a specified .obs column and optionally ordering within each group.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): An AnnData object containing the single cell data.</span>
<span class="sd">        groupby (str): The name of the column in .obs used for grouping cells in the heatmap.</span>
<span class="sd">        vmin (float, optional): Minimum value for color scaling. Defaults to 0.</span>
<span class="sd">        vmax (float, optional): Maximum value for color scaling. Defaults to 1.</span>
<span class="sd">        obsm_name (str, optional): The key in .obsm to retrieve the matrix for plotting. Defaults to &#39;SCN_score&#39;.</span>
<span class="sd">        order_by (str, optional): The name of the column in .obs used for ordering cells within each group. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: The function plots a heatmap and does not return any value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create a temporary AnnData object with the scores matrix and all original observations</span>
    <span class="n">adTemp</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">obsm_name</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>

    <span class="c1"># Determine sorting criteria</span>
    <span class="k">if</span> <span class="n">order_by</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sort_criteria</span> <span class="o">=</span> <span class="p">[</span><span class="n">groupby</span><span class="p">,</span> <span class="n">order_by</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sort_criteria</span> <span class="o">=</span> <span class="p">[</span><span class="n">groupby</span><span class="p">]</span>

    <span class="c1"># Determine the order of cells by sorting based on the criteria</span>
    <span class="n">sorted_order</span> <span class="o">=</span> <span class="n">adTemp</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">sort_criteria</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>

    <span class="c1"># Reorder adTemp according to the sorted order</span>
    <span class="n">adTemp</span> <span class="o">=</span> <span class="n">adTemp</span><span class="p">[</span><span class="n">sorted_order</span><span class="p">,</span> <span class="p">:]</span>

    <span class="c1"># Set figure dimensions and subplot adjustments</span>
    <span class="c1"># fsize = [5, 6]</span>
    <span class="c1"># plt.rcParams[&#39;figure.subplot.bottom&#39;] = figure_subplot_bottom</span>

    <span class="c1"># Plot the heatmap with the sorted and grouped data</span>
    <span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s1">&#39;figure.subplot.bottom&#39;</span><span class="p">:</span> <span class="n">figure_subplot_bottom</span><span class="p">}):</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">adTemp</span><span class="p">,</span> <span class="n">adTemp</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">Batlow_20</span><span class="o">.</span><span class="n">mpl_colormap</span><span class="p">,</span><span class="n">dendrogram</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">swap_axes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.make_bivariate_cmap" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">make_bivariate_cmap</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">make_bivariate_cmap</span><span class="p">(</span><span class="n">c00</span><span class="o">=</span><span class="s1">&#39;#f0f0f0&#39;</span><span class="p">,</span> <span class="n">c10</span><span class="o">=</span><span class="s1">&#39;#e31a1c&#39;</span><span class="p">,</span> <span class="n">c01</span><span class="o">=</span><span class="s1">&#39;#1f78b4&#39;</span><span class="p">,</span> <span class="n">c11</span><span class="o">=</span><span class="s1">&#39;#ffff00&#39;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">128</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Create a bivariate colormap by bilinear‐interpolating four corner colors.</p>
<p>This builds an (n × n) grid of RGB colors, blending smoothly between
the specified corner colors:
  - c00 at (low, low)
  - c10 at (high, low)
  - c01 at (low, high)
  - c11 at (high, high)</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>c00</code></b>
              (<code>str</code>, default:
                  <code>&#39;#f0f0f0&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Matplotlib color spec (hex, name, or RGB tuple) for the low/low corner.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>c10</code></b>
              (<code>str</code>, default:
                  <code>&#39;#e31a1c&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Color for the high/low corner.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>c01</code></b>
              (<code>str</code>, default:
                  <code>&#39;#1f78b4&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Color for the low/high corner.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>c11</code></b>
              (<code>str</code>, default:
                  <code>&#39;#ffff00&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Color for the high/high corner.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>n</code></b>
              (<code>int</code>, default:
                  <code>128</code>
)
          –
          <div class="doc-md-description">
            <p>Resolution per axis. The total length of the returned colormap is n*n.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
<b><code>ListedColormap</code></b> (              <code><span title="matplotlib.colors.ListedColormap">ListedColormap</span></code>
)          –
          <div class="doc-md-description">
            <p>A colormap with n*n entries blending between the four corners.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/helpers.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">make_bivariate_cmap</span><span class="p">(</span>
    <span class="n">c00</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;#f0f0f0&quot;</span><span class="p">,</span>
    <span class="n">c10</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;#e31a1c&quot;</span><span class="p">,</span>
    <span class="n">c01</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;#1f78b4&quot;</span><span class="p">,</span>
    <span class="n">c11</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;#ffff00&quot;</span><span class="p">,</span>
    <span class="n">n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">128</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ListedColormap</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a bivariate colormap by bilinear‐interpolating four corner colors.</span>

<span class="sd">    This builds an (n × n) grid of RGB colors, blending smoothly between</span>
<span class="sd">    the specified corner colors:</span>
<span class="sd">      - c00 at (low, low)</span>
<span class="sd">      - c10 at (high, low)</span>
<span class="sd">      - c01 at (low, high)</span>
<span class="sd">      - c11 at (high, high)</span>

<span class="sd">    Args:</span>
<span class="sd">        c00: Matplotlib color spec (hex, name, or RGB tuple) for the low/low corner.</span>
<span class="sd">        c10: Color for the high/low corner.</span>
<span class="sd">        c01: Color for the low/high corner.</span>
<span class="sd">        c11: Color for the high/high corner.</span>
<span class="sd">        n:   Resolution per axis. The total length of the returned colormap is n*n.</span>

<span class="sd">    Returns:</span>
<span class="sd">        ListedColormap: A colormap with n*n entries blending between the four corners.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Convert corner colors to RGB arrays</span>
    <span class="n">corners</span> <span class="o">=</span> <span class="p">{</span>
        <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">):</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">to_rgb</span><span class="p">(</span><span class="n">c00</span><span class="p">)),</span>
        <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">):</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">to_rgb</span><span class="p">(</span><span class="n">c10</span><span class="p">)),</span>
        <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">to_rgb</span><span class="p">(</span><span class="n">c01</span><span class="p">)),</span>
        <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">to_rgb</span><span class="p">(</span><span class="n">c11</span><span class="p">)),</span>
    <span class="p">}</span>

    <span class="c1"># Build an (n, n, 3) grid by bilinear interpolation</span>
    <span class="n">lut</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
    <span class="n">ys</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ys</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">xs</span><span class="p">):</span>
            <span class="n">lut</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">corners</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">y</span><span class="p">)</span> <span class="o">+</span>
                <span class="n">corners</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">*</span> <span class="n">x</span>       <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">y</span><span class="p">)</span> <span class="o">+</span>
                <span class="n">corners</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span> <span class="o">*</span> <span class="n">y</span>       <span class="o">+</span>
                <span class="n">corners</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)]</span> <span class="o">*</span> <span class="n">x</span>       <span class="o">*</span> <span class="n">y</span>
            <span class="p">)</span>

    <span class="c1"># Flatten to (n*n, 3) and return as a ListedColormap</span>
    <span class="k">return</span> <span class="n">ListedColormap</span><span class="p">(</span><span class="n">lut</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="n">n</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.scatter_genes_oneper" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">scatter_genes_oneper</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">scatter_genes_oneper</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">embedding_key</span><span class="o">=</span><span class="s1">&#39;X_spatial&#39;</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">99.5</span><span class="p">),</span> <span class="n">log_transform</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Reds&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">panel_width</span><span class="o">=</span><span class="mf">4.0</span><span class="p">,</span> <span class="n">n_rows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot expression of multiple genes on a 2D embedding arranged in a grid.</p>
<p>Each gene is optionally log-transformed, percentile-clipped, and rescaled to [0,1].
Cells are plotted on the embedding, colored by expression, with highest values
drawn on top. A single colorbar is placed to the right of the grid.
If <code>figsize</code> is None, each panel has width <code>panel_width</code> and height
proportional to the embedding's aspect ratio; total figure dims reflect
<code>n_rows</code> and computed columns.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>AnnData containing the embedding in <code>adata.obsm[embedding_key]</code>.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>embedding_key</code></b>
              (<code>str</code>, default:
                  <code>&#39;X_spatial&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Key in <code>.obsm</code> for an (n_obs, 2) coordinate array.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>genes</code></b>
              (<code><span title="typing.Sequence">Sequence</span>[str]</code>)
          –
          <div class="doc-md-description">
            <p>List of gene names to plot (must be in <code>adata.var_names</code>).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>spot_size</code></b>
              (<code>float</code>, default:
                  <code>2</code>
)
          –
          <div class="doc-md-description">
            <p>Marker size for scatter plots. Default 2.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>alpha</code></b>
              (<code>float</code>, default:
                  <code>0.9</code>
)
          –
          <div class="doc-md-description">
            <p>Transparency for markers. Default 0.9.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>clip_percentiles</code></b>
              (<code>tuple</code>, default:
                  <code>(0, 99.5)</code>
)
          –
          <div class="doc-md-description">
            <p>(low_pct, high_pct) to clip expression before rescaling.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>log_transform</code></b>
              (<code>bool</code>, default:
                  <code>True</code>
)
          –
          <div class="doc-md-description">
            <p>If True, apply <code>np.log1p</code> to raw expression.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>cmap</code></b>
              (<code><span title="typing.Union">Union</span>[str, <span title="matplotlib.pyplot.Colormap">Colormap</span>]</code>, default:
                  <code>&#39;Reds&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Colormap or name for all plots.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>figsize</code></b>
              (<code><span title="typing.Optional">Optional</span>[tuple]</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>(width, height) of entire figure. If None, computed from
<code>panel_width</code>, <code>n_rows</code>, and embedding aspect ratio.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>panel_width</code></b>
              (<code>float</code>, default:
                  <code>4.0</code>
)
          –
          <div class="doc-md-description">
            <p>Width (in inches) of each panel when <code>figsize</code> is None.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>n_rows</code></b>
              (<code>int</code>, default:
                  <code>1</code>
)
          –
          <div class="doc-md-description">
            <p>Number of rows in the grid. Default 1.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Raises:</span></p>
    <ul>
        <li class="doc-section-item field-body">
              <code>ValueError</code>
            –
          <div class="doc-md-description">
            <p>If embedding is missing/malformed or genes not found.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/spatial.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">scatter_genes_oneper</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">embedding_key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;X_spatial&quot;</span><span class="p">,</span>
    <span class="n">spot_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">alpha</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span>
    <span class="n">clip_percentiles</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">99.5</span><span class="p">),</span>
    <span class="n">log_transform</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">cmap</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">Colormap</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Reds&#39;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">panel_width</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">4.0</span><span class="p">,</span>
    <span class="n">n_rows</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot expression of multiple genes on a 2D embedding arranged in a grid.</span>

<span class="sd">    Each gene is optionally log-transformed, percentile-clipped, and rescaled to [0,1].</span>
<span class="sd">    Cells are plotted on the embedding, colored by expression, with highest values</span>
<span class="sd">    drawn on top. A single colorbar is placed to the right of the grid.</span>
<span class="sd">    If `figsize` is None, each panel has width `panel_width` and height</span>
<span class="sd">    proportional to the embedding&#39;s aspect ratio; total figure dims reflect</span>
<span class="sd">    `n_rows` and computed columns.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData containing the embedding in `adata.obsm[embedding_key]`.</span>
<span class="sd">        embedding_key: Key in `.obsm` for an (n_obs, 2) coordinate array.</span>
<span class="sd">        genes: List of gene names to plot (must be in `adata.var_names`).</span>
<span class="sd">        spot_size: Marker size for scatter plots. Default 2.</span>
<span class="sd">        alpha: Transparency for markers. Default 0.9.</span>
<span class="sd">        clip_percentiles: (low_pct, high_pct) to clip expression before rescaling.</span>
<span class="sd">        log_transform: If True, apply `np.log1p` to raw expression.</span>
<span class="sd">        cmap: Colormap or name for all plots.</span>
<span class="sd">        figsize: (width, height) of entire figure. If None, computed from</span>
<span class="sd">            `panel_width`, `n_rows`, and embedding aspect ratio.</span>
<span class="sd">        panel_width: Width (in inches) of each panel when `figsize` is None.</span>
<span class="sd">        n_rows: Number of rows in the grid. Default 1.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If embedding is missing/malformed or genes not found.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Helper to extract array</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_array</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">x</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s1">&#39;toarray&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="n">coords</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">embedding_key</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">coords</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">coords</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">2</span> <span class="ow">or</span> <span class="n">coords</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;adata.obsm[&#39;</span><span class="si">{</span><span class="n">embedding_key</span><span class="si">}</span><span class="s2">&#39;] must be an (n_obs, 2) array.&quot;</span><span class="p">)</span>
    <span class="n">x_vals</span><span class="p">,</span> <span class="n">y_vals</span> <span class="o">=</span> <span class="n">coords</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">coords</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>

    <span class="n">n_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">n_genes</span> <span class="o">/</span> <span class="n">n_rows</span><span class="p">)</span>
    <span class="c1"># Compute figsize if not provided</span>
    <span class="k">if</span> <span class="n">figsize</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">x_range</span> <span class="o">=</span> <span class="n">x_vals</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">x_vals</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
        <span class="n">y_range</span> <span class="o">=</span> <span class="n">y_vals</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">-</span> <span class="n">y_vals</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
        <span class="n">aspect</span> <span class="o">=</span> <span class="n">x_range</span> <span class="o">/</span> <span class="n">y_range</span> <span class="k">if</span> <span class="n">y_range</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mf">1.0</span>
        <span class="n">panel_height</span> <span class="o">=</span> <span class="n">panel_width</span> <span class="o">/</span> <span class="n">aspect</span>
        <span class="n">fig_width</span> <span class="o">=</span> <span class="n">panel_width</span> <span class="o">*</span> <span class="n">cols</span>
        <span class="n">fig_height</span> <span class="o">=</span> <span class="n">panel_height</span> <span class="o">*</span> <span class="n">n_rows</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fig_width</span><span class="p">,</span> <span class="n">fig_height</span> <span class="o">=</span> <span class="n">figsize</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">n_rows</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">fig_width</span><span class="p">,</span> <span class="n">fig_height</span><span class="p">),</span> <span class="n">squeeze</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">axes_flat</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="n">scatters</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genes</span><span class="p">):</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes_flat</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene &#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&#39; not found in adata.var_names.&quot;</span><span class="p">)</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">_get_array</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">log_transform</span><span class="p">:</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
        <span class="n">lo</span><span class="p">,</span> <span class="n">hi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="p">)</span>
        <span class="n">clipped</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="n">lo</span><span class="p">,</span> <span class="n">hi</span><span class="p">)</span>
        <span class="n">norm</span> <span class="o">=</span> <span class="p">(</span><span class="n">clipped</span> <span class="o">-</span> <span class="n">lo</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">hi</span> <span class="o">-</span> <span class="n">lo</span><span class="p">)</span> <span class="k">if</span> <span class="n">hi</span> <span class="o">&gt;</span> <span class="n">lo</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">clipped</span><span class="p">)</span>

        <span class="n">order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">norm</span><span class="p">)</span>
        <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">x_vals</span><span class="p">[</span><span class="n">order</span><span class="p">],</span>
            <span class="n">y_vals</span><span class="p">[</span><span class="n">order</span><span class="p">],</span>
            <span class="n">c</span><span class="o">=</span><span class="n">norm</span><span class="p">[</span><span class="n">order</span><span class="p">],</span>
            <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="n">spot_size</span><span class="p">,</span>
            <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span>
            <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([]);</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">scatters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sc</span><span class="p">)</span>

    <span class="c1"># Turn off unused axes</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span> <span class="n">n_rows</span><span class="o">*</span><span class="n">cols</span><span class="p">):</span>
        <span class="n">axes_flat</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>

    <span class="c1"># Adjust subplots to make room for colorbar</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">right</span><span class="o">=</span><span class="mf">0.85</span><span class="p">)</span>

    <span class="c1"># Colorbar axis on the right, spanning full height (15% margin)</span>
    <span class="n">cbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_axes</span><span class="p">([</span><span class="mf">0.88</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">])</span>
    <span class="n">cb</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">scatters</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">cax</span><span class="o">=</span><span class="n">cbar_ax</span><span class="p">)</span>
    <span class="n">cb</span><span class="o">.</span><span class="n">set_label</span><span class="p">(</span><span class="s1">&#39;normalized expression&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">(</span><span class="n">rect</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.85</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.scatter_qc_adata" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">scatter_qc_adata</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">scatter_qc_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">title_suffix</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Creates a figure with two scatter plot panels for visualizing data from an AnnData object.</p>
<p>The first panel shows 'total_counts' vs 'n_genes_by_counts', colored by 'pct_counts_mt'.
The second panel shows 'n_genes_by_counts' vs 'pct_counts_mt'. An optional title suffix
can be added to customize the axis titles.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>The AnnData object containing the dataset.
             Must contain 'total_counts', 'n_genes_by_counts', and 'pct_counts_mt' in <code>adata.obs</code>.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>title_suffix</code></b>
              (<code>str</code>, default:
                  <code>&#39;&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>A string to append to the axis titles, useful for specifying
                          experimental conditions (e.g., "C11 day 2"). Defaults to an empty string.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
<b><code>None</code></b>          –
          <div class="doc-md-description">
            <p>The function displays a matplotlib figure with two scatter plots.</p>
          </div>
        </li>
    </ul>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>plot_scatter_with_contours(adata, title_suffix="C11 day 2")</p>
</blockquote>
</blockquote>
</blockquote>
</details>
            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/scatter.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">scatter_qc_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">title_suffix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a figure with two scatter plot panels for visualizing data from an AnnData object.</span>

<span class="sd">    The first panel shows &#39;total_counts&#39; vs &#39;n_genes_by_counts&#39;, colored by &#39;pct_counts_mt&#39;.</span>
<span class="sd">    The second panel shows &#39;n_genes_by_counts&#39; vs &#39;pct_counts_mt&#39;. An optional title suffix</span>
<span class="sd">    can be added to customize the axis titles.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The AnnData object containing the dataset.</span>
<span class="sd">                         Must contain &#39;total_counts&#39;, &#39;n_genes_by_counts&#39;, and &#39;pct_counts_mt&#39; in `adata.obs`.</span>
<span class="sd">        title_suffix (str, optional): A string to append to the axis titles, useful for specifying</span>
<span class="sd">                                      experimental conditions (e.g., &quot;C11 day 2&quot;). Defaults to an empty string.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: The function displays a matplotlib figure with two scatter plots.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; plot_scatter_with_contours(adata, title_suffix=&quot;C11 day 2&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Extract necessary columns from the adata object</span>
    <span class="n">total_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">]</span>
    <span class="n">n_genes_by_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">]</span>
    <span class="n">pct_counts_mt</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pct_counts_mt&#39;</span><span class="p">]</span>

    <span class="c1"># Create a figure with two subplots (1 row, 2 columns)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

    <span class="c1"># First subplot: total_counts vs n_genes_by_counts, colored by pct_counts_mt</span>
    <span class="n">scatter1</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">total_counts</span><span class="p">,</span> <span class="n">n_genes_by_counts</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">pct_counts_mt</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Total Counts (</span><span class="si">{</span><span class="n">title_suffix</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Number of Genes by Counts (</span><span class="si">{</span><span class="n">title_suffix</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Total Counts vs Genes (</span><span class="si">{</span><span class="n">title_suffix</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="c1"># Add a colorbar</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">scatter1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;% Mito&#39;</span><span class="p">)</span>

    <span class="c1"># Second subplot: n_genes_by_counts vs pct_counts_mt</span>
    <span class="n">scatter2</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">n_genes_by_counts</span><span class="p">,</span> <span class="n">pct_counts_mt</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Number of Genes by Counts (</span><span class="si">{</span><span class="n">title_suffix</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;% Mito (</span><span class="si">{</span><span class="n">title_suffix</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Genes vs % Mito (</span><span class="si">{</span><span class="n">title_suffix</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>

    <span class="c1"># Adjust layout to avoid overlap</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

    <span class="c1"># Show the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.spatial_contours" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">spatial_contours</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">spatial_contours</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">spatial_key</span><span class="o">=</span><span class="s1">&#39;spatial&#39;</span><span class="p">,</span> <span class="n">summary_func</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">log_transform</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">99</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">contour_kwargs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">scatter_kwargs</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Scatter spatial expression of one or more genes with smooth contour overlay.</p>
<p>If multiple genes are provided, each is preprocessed (log1p → clip
→ normalize), then combined per cell via <code>summary_func</code> (e.g. mean, sum,
max) on the normalized values. A smooth contour of the summarized signal
is overlaid onto the spatial scatter.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>AnnData with spatial coordinates in <code>adata.obsm[spatial_key]</code>.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>genes</code></b>
              (<code><span title="typing.Union">Union</span>[str, <span title="typing.Sequence">Sequence</span>[str]]</code>)
          –
          <div class="doc-md-description">
            <p>Single gene name or list of gene names to plot (must be in <code>adata.var_names</code>).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>spatial_key</code></b>
              (<code>str</code>, default:
                  <code>&#39;spatial&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Key in <code>.obsm</code> for an (n_obs, 2) coords array.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>summary_func</code></b>
              (<code><span title="typing.Callable">Callable</span>[[<span title="numpy.ndarray">ndarray</span>], <span title="numpy.ndarray">ndarray</span>]</code>, default:
                  <code><span title="numpy.mean">mean</span></code>
)
          –
          <div class="doc-md-description">
            <p>Function to combine multiple normalized gene arrays
(takes an (n_obs, n_genes) array, returns length-n_obs array).
Defaults to <code>np.mean</code>.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>spot_size</code></b>
              (<code>float</code>, default:
                  <code>30</code>
)
          –
          <div class="doc-md-description">
            <p>Scatter marker size.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>alpha</code></b>
              (<code>float</code>, default:
                  <code>0.8</code>
)
          –
          <div class="doc-md-description">
            <p>Scatter alpha transparency.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>log_transform</code></b>
              (<code>bool</code>, default:
                  <code>True</code>
)
          –
          <div class="doc-md-description">
            <p>If True, apply <code>np.log1p</code> to raw expression before clipping.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>clip_percentiles</code></b>
              (<code>tuple</code>, default:
                  <code>(1, 99)</code>
)
          –
          <div class="doc-md-description">
            <p>Tuple <code>(low_pct, high_pct)</code> percentiles to clip each gene.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>cmap</code></b>
              (<code>str</code>, default:
                  <code>&#39;viridis&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Colormap name for the scatter (e.g. 'viridis').</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>contour_kwargs</code></b>
              (<code>dict</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>Dict of parameters for smoothing &amp; contouring:
- levels: int or list of levels (default 6)
- grid_res: int grid resolution (default 200)
- smooth_sigma: float Gaussian blur sigma (default 2)
- contour_kwargs: dict of line style kwargs (default {'colors':'k','linewidths':1})</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>scatter_kwargs</code></b>
              (<code>dict</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>Extra kwargs passed to <code>ax.scatter</code>.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Raises:</span></p>
    <ul>
        <li class="doc-section-item field-body">
              <code>ValueError</code>
            –
          <div class="doc-md-description">
            <p>If any gene is missing or spatial coords are malformed.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/spatial.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">spatial_contours</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">spatial_key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;spatial&#39;</span><span class="p">,</span>
    <span class="n">summary_func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">,</span>
    <span class="n">spot_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span>
    <span class="n">alpha</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span>
    <span class="n">log_transform</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">clip_percentiles</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">99</span><span class="p">),</span>
    <span class="n">cmap</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span>
    <span class="n">contour_kwargs</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">scatter_kwargs</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Scatter spatial expression of one or more genes with smooth contour overlay.</span>

<span class="sd">    If multiple genes are provided, each is preprocessed (log1p → clip</span>
<span class="sd">    → normalize), then combined per cell via `summary_func` (e.g. mean, sum,</span>
<span class="sd">    max) on the normalized values. A smooth contour of the summarized signal</span>
<span class="sd">    is overlaid onto the spatial scatter.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData with spatial coordinates in `adata.obsm[spatial_key]`.</span>
<span class="sd">        genes: Single gene name or list of gene names to plot (must be in `adata.var_names`).</span>
<span class="sd">        spatial_key: Key in `.obsm` for an (n_obs, 2) coords array.</span>
<span class="sd">        summary_func: Function to combine multiple normalized gene arrays</span>
<span class="sd">            (takes an (n_obs, n_genes) array, returns length-n_obs array).</span>
<span class="sd">            Defaults to `np.mean`.</span>
<span class="sd">        spot_size: Scatter marker size.</span>
<span class="sd">        alpha: Scatter alpha transparency.</span>
<span class="sd">        log_transform: If True, apply `np.log1p` to raw expression before clipping.</span>
<span class="sd">        clip_percentiles: Tuple `(low_pct, high_pct)` percentiles to clip each gene.</span>
<span class="sd">        cmap: Colormap name for the scatter (e.g. &#39;viridis&#39;).</span>
<span class="sd">        contour_kwargs: Dict of parameters for smoothing &amp; contouring:</span>
<span class="sd">            - levels: int or list of levels (default 6)</span>
<span class="sd">            - grid_res: int grid resolution (default 200)</span>
<span class="sd">            - smooth_sigma: float Gaussian blur sigma (default 2)</span>
<span class="sd">            - contour_kwargs: dict of line style kwargs (default {&#39;colors&#39;:&#39;k&#39;,&#39;linewidths&#39;:1})</span>
<span class="sd">        scatter_kwargs: Extra kwargs passed to `ax.scatter`.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If any gene is missing or spatial coords are malformed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># ensure genes is list</span>
    <span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">genes</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="nb">list</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">g</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene &#39;</span><span class="si">{</span><span class="n">g</span><span class="si">}</span><span class="s2">&#39; not found in adata.var_names.&quot;</span><span class="p">)</span>

    <span class="c1"># helper to extract numpy</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_array</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">x</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s1">&#39;toarray&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="c1"># preprocess each gene: extract, log1p, clip, normalize to [0,1]</span>
    <span class="n">normed</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">_get_array</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">g</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">log_transform</span><span class="p">:</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
        <span class="n">lo</span><span class="p">,</span> <span class="n">hi</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="p">)</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">vals</span><span class="p">,</span> <span class="n">lo</span><span class="p">,</span> <span class="n">hi</span><span class="p">)</span>
        <span class="n">normed</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">vals</span> <span class="o">-</span> <span class="n">lo</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">hi</span> <span class="o">-</span> <span class="n">lo</span><span class="p">)</span> <span class="k">if</span> <span class="n">hi</span> <span class="o">&gt;</span> <span class="n">lo</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">vals</span><span class="p">))</span>
    <span class="c1"># stack into (n_obs, n_genes)</span>
    <span class="n">M</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">(</span><span class="n">normed</span><span class="p">)</span>
    <span class="c1"># summarize across genes</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="n">summary_func</span><span class="p">(</span><span class="n">M</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># fetch spatial coords</span>
    <span class="n">coords</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">spatial_key</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">coords</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">coords</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;adata.obsm[&#39;</span><span class="si">{</span><span class="n">spatial_key</span><span class="si">}</span><span class="s2">&#39;] must be an (n_obs, 2) array.&quot;</span><span class="p">)</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">coords</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">coords</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>

    <span class="c1"># scatter</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sc_kw</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;c&quot;</span><span class="p">:</span> <span class="n">summary</span><span class="p">,</span> <span class="s2">&quot;cmap&quot;</span><span class="p">:</span> <span class="n">cmap</span><span class="p">,</span> <span class="s2">&quot;s&quot;</span><span class="p">:</span> <span class="n">spot_size</span><span class="p">,</span> <span class="s2">&quot;alpha&quot;</span><span class="p">:</span> <span class="n">alpha</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">scatter_kwargs</span><span class="p">:</span>
        <span class="n">sc_kw</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">scatter_kwargs</span><span class="p">)</span>
    <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="o">**</span><span class="n">sc_kw</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
    <span class="n">title</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gene_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes (</span><span class="si">{</span><span class="n">summary_func</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">)&quot;</span>
    <span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Spatial expression: </span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;x&#39;</span><span class="p">);</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;y&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sc</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;summarized (normalized)&quot;</span><span class="p">)</span>

    <span class="c1"># smooth + contour</span>
    <span class="c1"># default contour params</span>
    <span class="n">ck</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;levels&quot;</span><span class="p">:</span> <span class="mi">6</span><span class="p">,</span>
        <span class="s2">&quot;grid_res&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
        <span class="s2">&quot;smooth_sigma&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
        <span class="s2">&quot;contour_kwargs&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;colors&quot;</span><span class="p">:</span> <span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="s2">&quot;linewidths&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="n">contour_kwargs</span><span class="p">:</span>
        <span class="n">ck</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">contour_kwargs</span><span class="p">)</span>
    <span class="n">_smooth_contour</span><span class="p">(</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">summary</span><span class="p">,</span>
        <span class="n">levels</span><span class="o">=</span><span class="n">ck</span><span class="p">[</span><span class="s2">&quot;levels&quot;</span><span class="p">],</span>
        <span class="n">grid_res</span><span class="o">=</span><span class="n">ck</span><span class="p">[</span><span class="s2">&quot;grid_res&quot;</span><span class="p">],</span>
        <span class="n">smooth_sigma</span><span class="o">=</span><span class="n">ck</span><span class="p">[</span><span class="s2">&quot;smooth_sigma&quot;</span><span class="p">],</span>
        <span class="n">contour_kwargs</span><span class="o">=</span><span class="n">ck</span><span class="p">[</span><span class="s2">&quot;contour_kwargs&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.spatial_two_genes" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">spatial_two_genes</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">spatial_two_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">,</span> <span class="n">cmap</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">spatial_key</span><span class="o">=</span><span class="s1">&#39;X_spatial&#39;</span><span class="p">,</span> <span class="n">log_transform</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">99.5</span><span class="p">),</span> <span class="n">priority_metric</span><span class="o">=</span><span class="s1">&#39;sum&#39;</span><span class="p">,</span> <span class="n">show_xcoords</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_ycoords</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_bbox</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot two‐gene spatial expression with a bivariate colormap.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>AnnData with spatial coords in <code>adata.obsm[spatial_key]</code>.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>gene1</code></b>
              (<code>str</code>)
          –
          <div class="doc-md-description">
            <p>First gene name (must be in <code>adata.var_names</code>).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>gene2</code></b>
              (<code>str</code>)
          –
          <div class="doc-md-description">
            <p>Second gene name.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>cmap</code></b>
              (<code><span title="matplotlib.colors.ListedColormap">ListedColormap</span></code>)
          –
          <div class="doc-md-description">
            <p>Bivariate colormap from <code>make_bivariate_cmap</code> (n×n LUT).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>spot_size</code></b>
              (<code>float</code>, default:
                  <code>2</code>
)
          –
          <div class="doc-md-description">
            <p>Scatter point size.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>alpha</code></b>
              (<code>float</code>, default:
                  <code>0.9</code>
)
          –
          <div class="doc-md-description">
            <p>Point alpha transparency.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>spatial_key</code></b>
              (<code>str</code>, default:
                  <code>&#39;X_spatial&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Key in <code>adata.obsm</code> for an (n_obs, 2) coords array.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>log_transform</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>If True, apply <code>np.log1p</code> to raw expression.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>clip_percentiles</code></b>
              (<code>tuple</code>, default:
                  <code>(0, 99.5)</code>
)
          –
          <div class="doc-md-description">
            <p>Tuple <code>(low_pct, high_pct)</code> to clip each gene.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>priority_metric</code></b>
              (<code>str</code>, default:
                  <code>&#39;sum&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Which metric to sort drawing order by:
- 'sum': u + v (default)
- 'gene1': u only
- 'gene2': v only</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>show_xcoords</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>Whether to display x-axis ticks and labels.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>show_ycoords</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>Whether to display y-axis ticks and labels.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>show_bbox</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>Whether to display the bounding box (spines).</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>show_legend</code></b>
              (<code>bool</code>, default:
                  <code>True</code>
)
          –
          <div class="doc-md-description">
            <p>Whether to display the legend/colorbar.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>width_ratios</code></b>
              (<code><span title="typing.Tuple">Tuple</span>[float, float]</code>, default:
                  <code>(10, 1)</code>
)
          –
          <div class="doc-md-description">
            <p>2‐tuple giving the relative widths of
      (scatter_panel, legend_panel). Defaults to (3,1).</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Raises:</span></p>
    <ul>
        <li class="doc-section-item field-body">
              <code>ValueError</code>
            –
          <div class="doc-md-description">
            <p>If spatial coords are missing/malformed or
        if <code>priority_metric</code> is invalid.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/spatial.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">spatial_two_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">gene1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cmap</span><span class="p">:</span> <span class="n">ListedColormap</span><span class="p">,</span>
    <span class="n">spot_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">alpha</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.9</span><span class="p">,</span>
    <span class="n">spatial_key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;X_spatial&#39;</span><span class="p">,</span>
    <span class="n">log_transform</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">clip_percentiles</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">99.5</span><span class="p">),</span>
    <span class="n">priority_metric</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;sum&#39;</span><span class="p">,</span>
    <span class="n">show_xcoords</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">show_ycoords</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">show_bbox</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">show_legend</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">width_ratios</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot two‐gene spatial expression with a bivariate colormap.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: AnnData with spatial coords in `adata.obsm[spatial_key]`.</span>
<span class="sd">        gene1: First gene name (must be in `adata.var_names`).</span>
<span class="sd">        gene2: Second gene name.</span>
<span class="sd">        cmap: Bivariate colormap from `make_bivariate_cmap` (n×n LUT).</span>
<span class="sd">        spot_size: Scatter point size.</span>
<span class="sd">        alpha: Point alpha transparency.</span>
<span class="sd">        spatial_key: Key in `adata.obsm` for an (n_obs, 2) coords array.</span>
<span class="sd">        log_transform: If True, apply `np.log1p` to raw expression.</span>
<span class="sd">        clip_percentiles: Tuple `(low_pct, high_pct)` to clip each gene.</span>
<span class="sd">        priority_metric: Which metric to sort drawing order by:</span>
<span class="sd">            - &#39;sum&#39;: u + v (default)</span>
<span class="sd">            - &#39;gene1&#39;: u only</span>
<span class="sd">            - &#39;gene2&#39;: v only</span>
<span class="sd">        show_xcoords: Whether to display x-axis ticks and labels.</span>
<span class="sd">        show_ycoords: Whether to display y-axis ticks and labels.</span>
<span class="sd">        show_bbox: Whether to display the bounding box (spines).</span>
<span class="sd">        show_legend: Whether to display the legend/colorbar.</span>
<span class="sd">        width_ratios: 2‐tuple giving the relative widths of</span>
<span class="sd">                  (scatter_panel, legend_panel). Defaults to (3,1).</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If spatial coords are missing/malformed or</span>
<span class="sd">                    if `priority_metric` is invalid.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># 1) extract raw arrays</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_array</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">x</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s1">&#39;toarray&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">X1</span> <span class="o">=</span> <span class="n">_get_array</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">gene1</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
    <span class="n">X2</span> <span class="o">=</span> <span class="n">_get_array</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">gene2</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>

    <span class="c1"># 2) optional log1p</span>
    <span class="k">if</span> <span class="n">log_transform</span><span class="p">:</span>
        <span class="n">X1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">X1</span><span class="p">)</span>
        <span class="n">X2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">X2</span><span class="p">)</span>

    <span class="c1"># 3) percentile‐clip</span>
    <span class="n">lo1</span><span class="p">,</span> <span class="n">hi1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">X1</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="p">)</span>
    <span class="n">lo2</span><span class="p">,</span> <span class="n">hi2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">X2</span><span class="p">,</span> <span class="n">clip_percentiles</span><span class="p">)</span>
    <span class="n">X1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">X1</span><span class="p">,</span> <span class="n">lo1</span><span class="p">,</span> <span class="n">hi1</span><span class="p">)</span>
    <span class="n">X2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">X2</span><span class="p">,</span> <span class="n">lo2</span><span class="p">,</span> <span class="n">hi2</span><span class="p">)</span>

    <span class="c1"># 4) normalize to [0,1]</span>
    <span class="n">u</span> <span class="o">=</span> <span class="p">(</span><span class="n">X1</span> <span class="o">-</span> <span class="n">lo1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">hi1</span> <span class="o">-</span> <span class="n">lo1</span><span class="p">)</span> <span class="k">if</span> <span class="n">hi1</span> <span class="o">&gt;</span> <span class="n">lo1</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">X1</span><span class="p">)</span>
    <span class="n">v</span> <span class="o">=</span> <span class="p">(</span><span class="n">X2</span> <span class="o">-</span> <span class="n">lo2</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">hi2</span> <span class="o">-</span> <span class="n">lo2</span><span class="p">)</span> <span class="k">if</span> <span class="n">hi2</span> <span class="o">&gt;</span> <span class="n">lo2</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">X2</span><span class="p">)</span>

    <span class="c1"># 5) prepare LUT</span>
    <span class="n">m</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cmap</span><span class="o">.</span><span class="n">colors</span><span class="p">)</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">m</span><span class="p">))</span>
    <span class="n">C</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cmap</span><span class="o">.</span><span class="n">colors</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>

    <span class="c1"># 6) bilinear interpolate per‐cell</span>
    <span class="n">gu</span> <span class="o">=</span> <span class="n">u</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">);</span> <span class="n">gv</span> <span class="o">=</span> <span class="n">v</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">i0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">gu</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">);</span> <span class="n">j0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">gv</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">i1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">i0</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">);</span> <span class="n">j1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">j0</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">du</span> <span class="o">=</span> <span class="n">gu</span> <span class="o">-</span> <span class="n">i0</span><span class="p">;</span> <span class="n">dv</span> <span class="o">=</span> <span class="n">gv</span> <span class="o">-</span> <span class="n">j0</span>

    <span class="n">wa</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">du</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">dv</span><span class="p">)</span>
    <span class="n">wb</span> <span class="o">=</span> <span class="n">du</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">dv</span><span class="p">)</span>
    <span class="n">wc</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">du</span><span class="p">)</span> <span class="o">*</span> <span class="n">dv</span>
    <span class="n">wd</span> <span class="o">=</span> <span class="n">du</span> <span class="o">*</span> <span class="n">dv</span>

    <span class="n">c00</span> <span class="o">=</span> <span class="n">C</span><span class="p">[</span><span class="n">j0</span><span class="p">,</span> <span class="n">i0</span><span class="p">];</span> <span class="n">c10</span> <span class="o">=</span> <span class="n">C</span><span class="p">[</span><span class="n">j0</span><span class="p">,</span> <span class="n">i1</span><span class="p">]</span>
    <span class="n">c01</span> <span class="o">=</span> <span class="n">C</span><span class="p">[</span><span class="n">j1</span><span class="p">,</span> <span class="n">i0</span><span class="p">];</span> <span class="n">c11</span> <span class="o">=</span> <span class="n">C</span><span class="p">[</span><span class="n">j1</span><span class="p">,</span> <span class="n">i1</span><span class="p">]</span>

    <span class="n">cols_rgb</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">c00</span> <span class="o">*</span> <span class="n">wa</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">+</span>
        <span class="n">c10</span> <span class="o">*</span> <span class="n">wb</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">+</span>
        <span class="n">c01</span> <span class="o">*</span> <span class="n">wc</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span> <span class="o">+</span>
        <span class="n">c11</span> <span class="o">*</span> <span class="n">wd</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">hex_colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">to_hex</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cols_rgb</span><span class="p">]</span>

    <span class="c1"># 7) determine draw order</span>
    <span class="k">if</span> <span class="n">priority_metric</span> <span class="o">==</span> <span class="s1">&#39;sum&#39;</span><span class="p">:</span>
        <span class="n">priority</span> <span class="o">=</span> <span class="n">u</span> <span class="o">+</span> <span class="n">v</span>
    <span class="k">elif</span> <span class="n">priority_metric</span> <span class="o">==</span> <span class="s1">&#39;gene1&#39;</span><span class="p">:</span>
        <span class="n">priority</span> <span class="o">=</span> <span class="n">u</span>
    <span class="k">elif</span> <span class="n">priority_metric</span> <span class="o">==</span> <span class="s1">&#39;gene2&#39;</span><span class="p">:</span>
        <span class="n">priority</span> <span class="o">=</span> <span class="n">v</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;priority_metric must be &#39;sum&#39;, &#39;gene1&#39;, or &#39;gene2&#39;&quot;</span><span class="p">)</span>
    <span class="n">order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">priority</span><span class="p">)</span>

    <span class="c1"># 8) fetch and sort coords/colors</span>
    <span class="n">coords</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">spatial_key</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">coords</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">coords</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;adata.obsm[&#39;</span><span class="si">{</span><span class="n">spatial_key</span><span class="si">}</span><span class="s2">&#39;] must be an (n_obs, 2) array&quot;</span><span class="p">)</span>
    <span class="n">coords_sorted</span> <span class="o">=</span> <span class="n">coords</span><span class="p">[</span><span class="n">order</span><span class="p">]</span>
    <span class="n">colors_sorted</span> <span class="o">=</span> <span class="p">[</span><span class="n">hex_colors</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">order</span><span class="p">]</span>

    <span class="c1"># 9) plot scatter + optional legend</span>
    <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax_sc</span><span class="p">,</span> <span class="n">ax_cb</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span>
        <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
        <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
        <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;width_ratios&#39;</span><span class="p">:</span> <span class="n">width_ratios</span><span class="p">,</span> <span class="s1">&#39;wspace&#39;</span><span class="p">:</span> <span class="mf">0.3</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">ax_sc</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">coords_sorted</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">],</span>
        <span class="n">coords_sorted</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">],</span>
        <span class="n">c</span><span class="o">=</span><span class="n">colors_sorted</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="n">spot_size</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span>
    <span class="p">)</span>
    <span class="n">ax_sc</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
    <span class="n">ax_sc</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene1</span><span class="si">}</span><span class="s2"> :: </span><span class="si">{</span><span class="n">gene2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># axis display options</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">show_xcoords</span><span class="p">:</span>
        <span class="n">ax_sc</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">show_ycoords</span><span class="p">:</span>
        <span class="n">ax_sc</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">show_bbox</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">spine</span> <span class="ow">in</span> <span class="n">ax_sc</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">spine</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># legend/colorbar</span>
    <span class="k">if</span> <span class="n">show_legend</span><span class="p">:</span>
        <span class="n">lut_img</span> <span class="o">=</span> <span class="n">C</span>  <span class="c1"># shape (n,n,3)</span>
        <span class="n">ax_cb</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">lut_img</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s1">&#39;lower&#39;</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="c1"># ax_cb.set_xlabel(f&quot;{gene1}\nlow → high&quot;)</span>
        <span class="c1"># ax_cb.set_ylabel(f&quot;{gene2}\nlow → high&quot;)</span>
        <span class="n">ax_cb</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">ax_cb</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">ax_cb</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]);</span> <span class="n">ax_cb</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="n">ax_cb</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax_cb</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.stackedbar_composition" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">stackedbar_composition</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">stackedbar_composition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="p">,</span> <span class="n">obs_column</span><span class="o">=</span><span class="s1">&#39;SCN_class&#39;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bar_width</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">color_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">order_by_similarity</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">similarity_metric</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">,</span> <span class="n">include_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">legend_rows</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plots a stacked bar chart of cell type proportions for a single AnnData object grouped by a specified column.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>An AnnData object.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>groupby</code></b>
              (<code>str</code>)
          –
          <div class="doc-md-description">
            <p>The column in <code>.obs</code> to group by.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>obs_column</code></b>
              (<code>str</code>, default:
                  <code>&#39;SCN_class&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The name of the <code>.obs</code> column to use for categories. Defaults to 'SCN_class'.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>labels</code></b>
              (<code>List[str]</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>Custom labels for each group to be displayed on the x-axis.
If not provided, the unique values of the groupby column will be used. The length of <code>labels</code> must match
the number of unique groups.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>bar_width</code></b>
              (<code>float</code>, default:
                  <code>0.75</code>
)
          –
          <div class="doc-md-description">
            <p>The width of the bars in the plot. Defaults to 0.75.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>color_dict</code></b>
              (<code>Dict[str, str]</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>A dictionary mapping categories to specific colors. If not provided,
default colors will be used.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>ax</code></b>
              (<code><span title="matplotlib.axes.Axes">Axes</span></code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>The axis to plot on. If not provided, a new figure and axis will be created.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>order_by_similarity</code></b>
              (<code>bool</code>, default:
                  <code>False</code>
)
          –
          <div class="doc-md-description">
            <p>Whether to order the bars by similarity in composition. Defaults to False.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>similarity_metric</code></b>
              (<code>str</code>, default:
                  <code>&#39;correlation&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The metric to use for similarity ordering. Defaults to 'correlation'.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>include_legend</code></b>
              (<code>bool</code>, default:
                  <code>True</code>
)
          –
          <div class="doc-md-description">
            <p>Whether to include a legend in the plot. Defaults to True.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>legend_rows</code></b>
              (<code>int</code>, default:
                  <code>10</code>
)
          –
          <div class="doc-md-description">
            <p>The number of rows in the legend. Defaults to 10.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Raises:</span></p>
    <ul>
        <li class="doc-section-item field-body">
              <code>ValueError</code>
            –
          <div class="doc-md-description">
            <p>If the length of <code>labels</code> does not match the number of unique groups.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">stackedbar_composition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">obs_column</span><span class="o">=</span><span class="s1">&#39;your_column_name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stackedbar_composition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">obs_column</span><span class="o">=</span><span class="s1">&#39;your_column_name&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">include_legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">legend_rows</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/bar.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">stackedbar_composition</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> 
    <span class="n">groupby</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> 
    <span class="n">obs_column</span><span class="o">=</span><span class="s1">&#39;SCN_class&#39;</span><span class="p">,</span> 
    <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
    <span class="n">bar_width</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span> 
    <span class="n">color_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
    <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">order_by_similarity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">similarity_metric</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;correlation&#39;</span><span class="p">,</span>
    <span class="n">include_legend</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">legend_rows</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a stacked bar chart of cell type proportions for a single AnnData object grouped by a specified column.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (anndata.AnnData): An AnnData object.</span>
<span class="sd">        groupby (str): The column in `.obs` to group by.</span>
<span class="sd">        obs_column (str, optional): The name of the `.obs` column to use for categories. Defaults to &#39;SCN_class&#39;.</span>
<span class="sd">        labels (List[str], optional): Custom labels for each group to be displayed on the x-axis.</span>
<span class="sd">            If not provided, the unique values of the groupby column will be used. The length of `labels` must match</span>
<span class="sd">            the number of unique groups.</span>
<span class="sd">        bar_width (float, optional): The width of the bars in the plot. Defaults to 0.75.</span>
<span class="sd">        color_dict (Dict[str, str], optional): A dictionary mapping categories to specific colors. If not provided,</span>
<span class="sd">            default colors will be used.</span>
<span class="sd">        ax (matplotlib.axes.Axes, optional): The axis to plot on. If not provided, a new figure and axis will be created.</span>
<span class="sd">        order_by_similarity (bool, optional): Whether to order the bars by similarity in composition. Defaults to False.</span>
<span class="sd">        similarity_metric (str, optional): The metric to use for similarity ordering. Defaults to &#39;correlation&#39;.</span>
<span class="sd">        include_legend (bool, optional): Whether to include a legend in the plot. Defaults to True.</span>
<span class="sd">        legend_rows (int, optional): The number of rows in the legend. Defaults to 10.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the length of `labels` does not match the number of unique groups.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; stackedbar_composition(adata, groupby=&#39;sample&#39;, obs_column=&#39;your_column_name&#39;)</span>
<span class="sd">        &gt;&gt;&gt; fig, ax = plt.subplots()</span>
<span class="sd">        &gt;&gt;&gt; stackedbar_composition(adata, groupby=&#39;sample&#39;, obs_column=&#39;your_column_name&#39;, ax=ax, include_legend=False, legend_rows=5)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Ensure the groupby column exists in .obs</span>
    <span class="k">if</span> <span class="n">groupby</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The groupby column &#39;</span><span class="si">{</span><span class="n">groupby</span><span class="si">}</span><span class="s2">&#39; does not exist in the .obs attribute.&quot;</span><span class="p">)</span>

    <span class="c1"># Check if groupby column is categorical or not</span>
    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">api</span><span class="o">.</span><span class="n">types</span><span class="o">.</span><span class="n">is_categorical_dtype</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]):</span>
        <span class="n">unique_groups</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">unique_groups</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Extract unique groups and ensure labels are provided or create default ones</span>
    <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="n">unique_groups</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_groups</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Length of &#39;labels&#39; must match the number of unique groups.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">color_dict</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">color_dict</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;SCN_class_colors&#39;</span><span class="p">,</span> <span class="p">{})</span>

    <span class="c1"># Extracting category proportions per group</span>
    <span class="n">category_counts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">categories</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">unique_groups</span><span class="p">:</span>
        <span class="n">subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby</span><span class="p">]</span> <span class="o">==</span> <span class="n">group</span><span class="p">]</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">subset</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_column</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">category_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
        <span class="n">categories</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="n">categories</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span>

    <span class="c1"># Preparing the data for plotting</span>
    <span class="n">proportions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">categories</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">unique_groups</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">counts</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">category_counts</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">category</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">j</span> <span class="o">=</span> <span class="n">categories</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">category</span><span class="p">)</span>
            <span class="n">proportions</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">category</span><span class="p">]</span>

    <span class="c1"># Ordering groups by similarity if requested</span>
    <span class="k">if</span> <span class="n">order_by_similarity</span><span class="p">:</span>
        <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">pdist</span><span class="p">(</span><span class="n">proportions</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="n">similarity_metric</span><span class="p">)</span>
        <span class="n">linkage_matrix</span> <span class="o">=</span> <span class="n">linkage</span><span class="p">(</span><span class="n">dist_matrix</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;average&#39;</span><span class="p">)</span>
        <span class="n">order</span> <span class="o">=</span> <span class="n">leaves_list</span><span class="p">(</span><span class="n">linkage_matrix</span><span class="p">)</span>
        <span class="n">proportions</span> <span class="o">=</span> <span class="n">proportions</span><span class="p">[:,</span> <span class="n">order</span><span class="p">]</span>
        <span class="n">unique_groups</span> <span class="o">=</span> <span class="p">[</span><span class="n">unique_groups</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">order</span><span class="p">]</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">labels</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">order</span><span class="p">]</span>

    <span class="c1"># Plotting</span>
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">figure</span>

    <span class="n">bottom</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_groups</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">category</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">categories</span><span class="p">):</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">color_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">category</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span>
            <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_groups</span><span class="p">)),</span> 
            <span class="n">proportions</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> 
            <span class="n">bottom</span><span class="o">=</span><span class="n">bottom</span><span class="p">,</span> 
            <span class="n">label</span><span class="o">=</span><span class="n">category</span><span class="p">,</span> 
            <span class="n">width</span><span class="o">=</span><span class="n">bar_width</span><span class="p">,</span> 
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> 
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">color</span>
        <span class="p">)</span>
        <span class="n">bottom</span> <span class="o">+=</span> <span class="n">proportions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_groups</span><span class="p">)))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Proportion&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">obs_column</span><span class="si">}</span><span class="s1"> proportions by </span><span class="si">{</span><span class="n">groupby</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">include_legend</span><span class="p">:</span>
        <span class="n">num_columns</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span> <span class="o">/</span> <span class="n">legend_rows</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Classes&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="n">num_columns</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">ax</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.stackedbar_composition_list" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">stackedbar_composition_list</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">stackedbar_composition_list</span><span class="p">(</span><span class="n">adata_list</span><span class="p">,</span> <span class="n">obs_column</span><span class="o">=</span><span class="s1">&#39;SCN_class&#39;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bar_width</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">color_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;outside center right&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plots a stacked bar chart of category proportions for a list of AnnData objects.</p>
<p>This function takes a list of AnnData objects, and for a specified column in the <code>.obs</code> attribute,
it plots a stacked bar chart. Each bar represents an AnnData object with segments showing the proportion
of each category within that object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata_list</code></b>
              (<code>List[<span title="anndata.AnnData">AnnData</span>]</code>)
          –
          <div class="doc-md-description">
            <p>A list of AnnData objects.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>obs_column</code></b>
              (<code>str</code>, default:
                  <code>&#39;SCN_class&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The name of the <code>.obs</code> column to use for categories. Defaults to 'SCN_class'.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>labels</code></b>
              (<code>List[str]</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>Custom labels for each AnnData object to be displayed on the x-axis.
If not provided, defaults to 'AnnData {i+1}' for each object. The length of <code>labels</code> must match
the number of AnnData objects provided.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>bar_width</code></b>
              (<code>float</code>, default:
                  <code>0.75</code>
)
          –
          <div class="doc-md-description">
            <p>The width of the bars in the plot. Defaults to 0.75.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>color_dict</code></b>
              (<code>Dict[str, str]</code>, default:
                  <code>None</code>
)
          –
          <div class="doc-md-description">
            <p>A dictionary mapping categories to specific colors. If not provided,
default colors will be used.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Raises:</span></p>
    <ul>
        <li class="doc-section-item field-body">
              <code>ValueError</code>
            –
          <div class="doc-md-description">
            <p>If the length of <code>labels</code> does not match the number of AnnData objects.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">plot_cell_type_proportions</span><span class="p">([</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">],</span> <span class="n">obs_column</span><span class="o">=</span><span class="s1">&#39;your_column_name&#39;</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Sample 1&#39;</span><span class="p">,</span> <span class="s1">&#39;Sample 2&#39;</span><span class="p">])</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/bar.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">stackedbar_composition_list</span><span class="p">(</span>
    <span class="n">adata_list</span><span class="p">,</span> 
    <span class="n">obs_column</span> <span class="o">=</span> <span class="s1">&#39;SCN_class&#39;</span><span class="p">,</span> 
    <span class="n">labels</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> 
    <span class="n">bar_width</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span> 
    <span class="n">color_dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">legend_loc</span> <span class="o">=</span> <span class="s2">&quot;outside center right&quot;</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots a stacked bar chart of category proportions for a list of AnnData objects.</span>

<span class="sd">    This function takes a list of AnnData objects, and for a specified column in the `.obs` attribute,</span>
<span class="sd">    it plots a stacked bar chart. Each bar represents an AnnData object with segments showing the proportion</span>
<span class="sd">    of each category within that object.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata_list (List[anndata.AnnData]): A list of AnnData objects.</span>
<span class="sd">        obs_column (str, optional): The name of the `.obs` column to use for categories. Defaults to &#39;SCN_class&#39;.</span>
<span class="sd">        labels (List[str], optional): Custom labels for each AnnData object to be displayed on the x-axis.</span>
<span class="sd">            If not provided, defaults to &#39;AnnData {i+1}&#39; for each object. The length of `labels` must match</span>
<span class="sd">            the number of AnnData objects provided.</span>
<span class="sd">        bar_width (float, optional): The width of the bars in the plot. Defaults to 0.75.</span>
<span class="sd">        color_dict (Dict[str, str], optional): A dictionary mapping categories to specific colors. If not provided,</span>
<span class="sd">            default colors will be used.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the length of `labels` does not match the number of AnnData objects.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; plot_cell_type_proportions([adata1, adata2], obs_column=&#39;your_column_name&#39;, labels=[&#39;Sample 1&#39;, &#39;Sample 2&#39;])</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Ensure labels are provided, or create default ones</span>
    <span class="k">if</span> <span class="n">labels</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;AnnData </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_list</span><span class="p">))]</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">adata_list</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Length of &#39;labels&#39; must match the number of AnnData objects provided.&quot;</span><span class="p">)</span>

    <span class="c1"># Extracting category proportions</span>
    <span class="n">category_counts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">categories</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">adata_list</span><span class="p">:</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">obs_column</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">category_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
        <span class="n">categories</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="n">categories</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span>

    <span class="c1"># Preparing the data for plotting</span>
    <span class="n">proportions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">categories</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">adata_list</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">counts</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">category_counts</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">category</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">j</span> <span class="o">=</span> <span class="n">categories</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">category</span><span class="p">)</span>
            <span class="n">proportions</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">counts</span><span class="p">[</span><span class="n">category</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">color_dict</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">color_dict</span> <span class="o">=</span> <span class="n">adata_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;SCN_class_colors&#39;</span><span class="p">]</span> <span class="c1"># should parameterize this</span>

    <span class="c1"># Plotting</span>
    <span class="c1">#### fig, ax = plt.subplots()</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">constrained_layout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">bottom</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_list</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">category</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">categories</span><span class="p">):</span>
        <span class="n">color</span> <span class="o">=</span> <span class="n">color_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span> <span class="k">if</span> <span class="n">color_dict</span> <span class="ow">and</span> <span class="n">category</span> <span class="ow">in</span> <span class="n">color_dict</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span>
            <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_list</span><span class="p">)),</span> 
            <span class="n">proportions</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> 
            <span class="n">bottom</span><span class="o">=</span><span class="n">bottom</span><span class="p">,</span> 
            <span class="n">label</span><span class="o">=</span><span class="n">category</span><span class="p">,</span> 
            <span class="n">width</span><span class="o">=</span><span class="n">bar_width</span><span class="p">,</span> 
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> 
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">color</span>
        <span class="p">)</span>
        <span class="n">bottom</span> <span class="o">+=</span> <span class="n">proportions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_list</span><span class="p">)))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Proportion&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">obs_column</span><span class="si">}</span><span class="s1"> proportions&#39;</span><span class="p">)</span>
    <span class="c1"># ax.legend(title=&#39;Classes&#39;, bbox_to_anchor=(1.05, 1), loc=&#39;upper left&#39;)</span>
    <span class="c1">## legend = fig.legend(title=&#39;Classes&#39;, loc=&quot;outside right upper&quot;, frameon=False)#, bbox_to_anchor=(1.05, 1), loc=&#39;upper left&#39;)</span>
    <span class="c1">## legend_height = legend.get_window_extent().height / fig.dpi  # in inches</span>

    <span class="c1"># Add legend</span>
    <span class="n">legend_handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">)</span> <span class="k">for</span> <span class="n">label</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="n">color_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()]</span>
    <span class="c1"># legend = ax.legend(handles=legend_handles, bbox_to_anchor=(1.05, 1), loc=&#39;upper left&#39;, frameon=False)</span>
    <span class="c1">##### legend = fig.legend(title=&#39;Classes&#39;, loc=&quot;outside right upper&quot;, frameon=False)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">legend_handles</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1">##### legend_height = legend.get_window_extent().height / fig.dpi  # in inches</span>

    <span class="c1"># fig_height = fig.get_size_inches()[1]  # current height in inches</span>
    <span class="c1">#### fig.set_size_inches(fig.get_size_inches()[0], legend_height )</span>
    <span class="c1"># plt.tight_layout()</span>
    <span class="c1"># plt.show()</span>
    <span class="k">return</span> <span class="n">fig</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.umap_scores" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">umap_scores</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">umap_scores</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">scn_classes</span><span class="p">,</span> <span class="n">obsm_name</span><span class="o">=</span><span class="s1">&#39;SCN_score&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">display</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plots UMAP projections of scRNA-seq data with specified scores.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>The AnnData object containing the scRNA-seq data.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>scn_classes</code></b>
              (<code>list</code>)
          –
          <div class="doc-md-description">
            <p>A list of SCN classes to visualize on the UMAP.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>obsm_name</code></b>
              (<code>str</code>, default:
                  <code>&#39;SCN_score&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>The name of the obsm key containing the SCN scores. Defaults to 'SCN_score'.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>alpha</code></b>
              (<code>float</code>, default:
                  <code>0.75</code>
)
          –
          <div class="doc-md-description">
            <p>The transparency level of the points on the UMAP plot. Defaults to 0.75.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>s</code></b>
              (<code>int</code>, default:
                  <code>10</code>
)
          –
          <div class="doc-md-description">
            <p>The size of the points on the UMAP plot. Defaults to 10.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>display</code></b>
              (<code>bool</code>, default:
                  <code>True</code>
)
          –
          <div class="doc-md-description">
            <p>If True, the plot is displayed immediately. If False, the axis object is returned. Defaults to True.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
          –
          <div class="doc-md-description">
            <p>matplotlib.axes.Axes or None: 
If <code>display</code> is False, returns the matplotlib axes object. Otherwise, returns None.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/dot.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">umap_scores</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="n">scn_classes</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
    <span class="n">obsm_name</span><span class="o">=</span><span class="s1">&#39;SCN_score&#39;</span><span class="p">,</span>
    <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span>
    <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">display</span><span class="o">=</span><span class="kc">True</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots UMAP projections of scRNA-seq data with specified scores.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): </span>
<span class="sd">            The AnnData object containing the scRNA-seq data.</span>
<span class="sd">        scn_classes (list): </span>
<span class="sd">            A list of SCN classes to visualize on the UMAP.</span>
<span class="sd">        obsm_name (str, optional): </span>
<span class="sd">            The name of the obsm key containing the SCN scores. Defaults to &#39;SCN_score&#39;.</span>
<span class="sd">        alpha (float, optional): </span>
<span class="sd">            The transparency level of the points on the UMAP plot. Defaults to 0.75.</span>
<span class="sd">        s (int, optional): </span>
<span class="sd">            The size of the points on the UMAP plot. Defaults to 10.</span>
<span class="sd">        display (bool, optional): </span>
<span class="sd">            If True, the plot is displayed immediately. If False, the axis object is returned. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        matplotlib.axes.Axes or None: </span>
<span class="sd">            If `display` is False, returns the matplotlib axes object. Otherwise, returns None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create a temporary AnnData object with the desired obsm</span>
    <span class="n">adTemp</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">obsm_name</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
    <span class="n">adTemp</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Create the UMAP plot</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adTemp</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">scn_classes</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Display or return the axis</span>
    <span class="k">if</span> <span class="n">display</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">ax</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="pySingleCellNet.plotting.umi_counts_ranked" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">umi_counts_ranked</span>


</h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">umi_counts_ranked</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">total_counts_column</span><span class="o">=</span><span class="s1">&#39;total_counts&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Identifies and plors the knee point of the UMI count distribution in an AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <ul>
        <li class="doc-section-item field-body">
            <b><code>adata</code></b>
              (<code><span title="anndata.AnnData">AnnData</span></code>)
          –
          <div class="doc-md-description">
            <p>The input AnnData object.</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>total_counts_column</code></b>
              (<code>str</code>, default:
                  <code>&#39;total_counts&#39;</code>
)
          –
          <div class="doc-md-description">
            <p>Column in <code>adata.obs</code> containing total UMI counts. Default is "total_counts".</p>
          </div>
        </li>
        <li class="doc-section-item field-body">
            <b><code>show</code></b>
              (<code>bool</code>)
          –
          <div class="doc-md-description">
            <p>If True, displays a log-log plot with the knee point. Default is True.</p>
          </div>
        </li>
    </ul>


<p><span class="doc-section-title">Returns:</span></p>
    <ul>
        <li class="doc-section-item field-body">
<b><code>float</code></b>          –
          <div class="doc-md-description">
            <p>The UMI count value at the knee point.</p>
          </div>
        </li>
    </ul>

            <details class="quote">
              <summary>Source code in <code>src/pySingleCellNet/plotting/dot.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">umi_counts_ranked</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">total_counts_column</span><span class="o">=</span><span class="s2">&quot;total_counts&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identifies and plors the knee point of the UMI count distribution in an AnnData object.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        adata (AnnData): The input AnnData object.</span>
<span class="sd">        total_counts_column (str): Column in `adata.obs` containing total UMI counts. Default is &quot;total_counts&quot;.</span>
<span class="sd">        show (bool): If True, displays a log-log plot with the knee point. Default is True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: The UMI count value at the knee point.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract total UMI counts</span>
    <span class="n">umi_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">total_counts_column</span><span class="p">]</span>

    <span class="c1"># Sort UMI counts in descending order</span>
    <span class="n">sorted_umi_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">umi_counts</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Compute cumulative UMI counts (normalized to a fraction)</span>
    <span class="n">cumulative_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">sorted_umi_counts</span><span class="p">)</span>
    <span class="n">cumulative_fraction</span> <span class="o">=</span> <span class="n">cumulative_counts</span> <span class="o">/</span> <span class="n">cumulative_counts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Compute derivatives to identify the knee point</span>
    <span class="n">first_derivative</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">gradient</span><span class="p">(</span><span class="n">cumulative_fraction</span><span class="p">)</span>
    <span class="n">second_derivative</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">gradient</span><span class="p">(</span><span class="n">first_derivative</span><span class="p">)</span>

    <span class="c1"># Find the index of the maximum curvature (knee point)</span>
    <span class="n">knee_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">second_derivative</span><span class="p">)</span>
    <span class="n">knee_point_value</span> <span class="o">=</span> <span class="n">sorted_umi_counts</span><span class="p">[</span><span class="n">knee_idx</span><span class="p">]</span>

    <span class="c1"># Generate log-log plot</span>
    <span class="n">cell_ranks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sorted_umi_counts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cell_ranks</span><span class="p">,</span> <span class="n">sorted_umi_counts</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;UMI Counts&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">cell_ranks</span><span class="p">[</span><span class="n">knee_idx</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Knee Point: </span><span class="si">{</span><span class="n">knee_point_value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;UMI Counts Per Cell (Log-Log Scale)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Cell Rank (Descending)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Total UMI Counts&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2024 Cahan Lab
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": ["content.footnote.tooltips", "navigation.expand", "toc.integrate"], "search": "assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.88dd0f4e.min.js"></script>
      
        <script src="javascripts/extra.js"></script>
      
    
  </body>
</html>